apps/system/js/advanced_telemetry.js

```diff
@@ -0,0 +1,426 @@
+        else {
+          loginfo('Stop Collecting');
+          this.stopCollecting();
+        }
+      }.bind(this);
+
+    SettingsListener.observe(AT.TELEMETRY_ENABLED_KEY,
+      false, this.advancedTelemetryEnabledListener);
+  };
+
+  AT.prototype.startCollecting = function startCollecting(done) {
+    var self = this;
+    // If we're already running there is nothing to start
+    if (this.collecting) {
+      return;
+    }
+    this.collecting = true;
+
+    this.metrics = new BatchTiming(true);
+
+    getDeviceID();
+
+    function getDeviceID() {
+      asyncStorage.getItem(DEVICE_ID_KEY, function(value) {
+        if (value) {
+          self.deviceID = value;
+        }
+        else {
+          // Our device id does not need to be unique, just probably unique.
+          self.deviceID = uuid();
+          asyncStorage.setItem(DEVICE_ID_KEY, self.deviceID);
+        }
+
+        // Move on to the next step in the startup process
+        registerHandlers();
+      });
+    }
+
+    function registerHandlers() {
+      // Basic event handlers
+      EVENT_TYPES.forEach(function (type) {
+        window.addEventListener(type, self);
+      });
+
+      self.startBatch();
+
+      if (done) {
+        loginfo('Initialization Complete');
+        done();
+      }
+    }
+  };
+
+  // This method shuts everything down and is only exposed for unit testing.
+  // Note that this is not the same as the stopCollecting() method which is
+  // used to stop data collection but keep the module running.
+  AT.prototype.stop = function stop() {
+
+  };
+
+  // Reset (or initialize) the AdvancedTelemetry instance variables
+  AT.prototype.reset = function() {
+    this.collecting = false;
+    this.metrics = null;
+    this.deviceID = null;
+  };
+
+  AT.prototype.handleEvent = function handleEvent(e) {
+    switch (e.type) {
+      case HUDMETRICS:
+        this.transmit(e.detail);
+        break;
+      default:
+        break;
+    }
+  };
+
+  // Signal gecko to send the histograms with a telemetry message.
+  AT.prototype.getPayload = function getPayload() {
+    console.info('telemetry|MGMT|TIMETOSHIP');
+  };
+
+  // Start a timer to notify when to send the payload.
+  AT.prototype.startBatch = function startBatch() {
+    var self = this;
+    this.timer = setTimeout(function() {
+      if (navigator.onLine) {
+        self.getPayload();
+      }
+    }, self.metrics.getInterval());
+  };
+
+  AT.prototype.stopBatch = function stopBatch() {
+    clearTimeout(this.timer);
+  };
+
+  // Once we have the histogram payload, we can send it to the telemetry
+  // server.  Once we receive 200 OK back from telemetry server, we can
+  // start a fresh batch of metrics.
+  AT.prototype.transmit = function transmit(payload) {
+    var self = this;
+
+    if (!this.collecting) {
+      return;
+    }
+
+    // But assume that it will succeed and start collecting new metrics now
+    this.metrics = new BatchTiming(false);
+    this.startBatch();
```

Comment:
it looks like you call `startBatch()` any time the `advanced-telemetry-update` event is sent, but I'm unclear on why you need to send the encoded string `"'telemetry|MGMT|TIMETOSHIP'"` after a timeout from this function. Can you not just do it synchronously?